### PR TITLE
MM-28678 Cypress tests on CircleCI

### PR DESCRIPTION
**Pull request by cpoile**
#### Summary
- All the cypress tests will work on CircleCI now
- Had to make a small change to the tests -- the selector we were using for latest post was flaky. Turns out it doesn't work very well when there's only one post in the channel. So make two posts and then try to click on the post menu. Now it's more robust.
- Made some changes to the CircleCI caching.
- I'm not processing or posting the Cypress report yet. Usually people send it to an S3 bucket and link it in a webhook event. I'm not sure if we want to do that and send it to the EE: IR channel? Might get a bit noisy. It would require more mana to hook reporting up, and I thought getting the tests to run was more impactful.
- Shout out to @saturninoabril for the e2e circleci template -- that saved a lot of time!

#### Ticket Link
- https://mattermost.atlassian.net/browse/MM-28678

### CODE REVIEW

**Comment by lieut-data** [comment]
🤯

**Review by agarciamontoro — CHANGES_REQUESTED** [review]
This is amazing, thank you! :pray: 

I see that the data in test_data.sql is from a dump, but I'm curious where that data comes from.

**Review by lieut-data — APPROVED** [review]
:+1: to @agarciamontoro's feedback, but otherwise LGTM.

File: .circleci/config.yml [hunk 1]
```diff
@@ -136,19 +155,115 @@ jobs:
...
+    steps:
+      - setup_remote_docker
+      - run:
+          name: Wait for Inbucket
+          command: |
+            until curl --max-time 5 --output - localhost:10080; do echo waiting for Inbucket; sleep 5; done;
+      - run:
+          name: Wait for Elasticsearch
+          command: |
+            until curl --max-time 5 --output - localhost:9200; do echo waiting for Elasticsearch; sleep 5; done;
```
[the remark this answers is not in the document]
Comment (cpoile) [reply .circleci/config.yml:241]: I'm hesitant to change it -- it's working, and this part is originally Saturn's. I don't mind if we wait a bit too long -- better than not waiting long enough. :)

File: server/sqlstore/data_migration.go [hunk 2]
```diff
@@ -230,8 +233,10 @@ func DataMigration(store *SQLStore, tx *sqlx.Tx, kvAPI KVAPI) error {
 			return errors.Wrapf(err, "failed inserting data into Playbook table")
 		}
 
-		if _, err := store.execBuilder(tx, playbookMemberInsert); err != nil {
-			return errors.Wrapf(err, "failed inserting data into PlaybookMember table")
+		if okToInsertMembers {
+			if _, err := store.execBuilder(tx, playbookMemberInsert); err != nil {
+				return errors.Wrapf(err, "failed inserting data into PlaybookMember table")
+			}
```
[the remark this answers is not in the document]
Comment (cpoile) [reply server/sqlstore/data_migration.go:239]: Yep, that's a good idea! Done!

**Comment by cpoile** [comment]
The dump comes from me running a fresh postgres install and running `make test-data`. Just to make sure we know exactly what data we're going to have.

**Comment by cpoile** [comment]
@agarciamontoro Please merge this if you're happy with the changes!

File: tests-e2e/cypress/support/plugin_ui_commands.js [hunk 3]
```diff
@@ -62,8 +62,11 @@ Cypress.Commands.add('startIncidentFromRHS', (playbookName, incidentName) => {
 // Starts incident from the post menu
 // function startIncidentFromPostMenu(incidentName) {
 Cypress.Commands.add('startIncidentFromPostMenu', (playbookName, incidentName) => {
-    //post a message as user to avoid system message
+    // post a message as user to avoid system message
     cy.findByTestId('post_textbox').clear().type('new message here{enter}');
+
+    // post a second message because cypress has trouble finding latest post when there's only one message
+    cy.findByTestId('post_textbox').clear().type('another new message here{enter}');
```
Comment (prapti) [inline tests-e2e/cypress/support/plugin_ui_commands.js:69]: It's strange and I don't know why this is needed, but all the tests are passing now so I don't want to dig into it right at this moment. I'll take a note of this for future cleanup.

**Review by prapti — APPROVED** [review]
Looks great! I tested the files where changes in `plugin_ui_command` would be used, and they all passed.
Thank you for doing this @cpoile!!

**Comment by prapti** [comment]
And just realized that I reviewed before @agarciamontoro finished his review. But if I need to review it again, please let me know :)

File: .circleci/config.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (agarciamontoro) [reply .circleci/config.yml:240]: Sounds good!

File: server/sqlstore/data_migration.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (agarciamontoro) [reply server/sqlstore/data_migration.go]: Thanks :)

**Review by agarciamontoro — APPROVED** [review]

File: tests-e2e/cypress/support/plugin_ui_commands.js (hunk 3 above)
Comment (cpoile) [reply tests-e2e/cypress/support/plugin_ui_commands.js:69]: @prapti I think it has something to do with the selector for last message -- if there's only one message in the channel,  that message won't have the exact format of id that the selector is looking for, so it just waits and waits and then fails.  But I could be wrong, please let me know what you uncover!

**Pull request merged**